File: app/src/main/java/com/example/weatherapp1/WeatherList/MainActivity.java
package com.example.weatherapp1.WeatherList;

import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.NavUtils;
import android.support.v7.app.AppCompatActivity;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.ListView;
import android.widget.TextView;

import com.example.weatherapp1.Data.Model.DisplayClass;
import com.example.weatherapp1.R;
import com.example.weatherapp1.WeatherManipulator.Main3Activity;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.Completable;
import io.reactivex.CompletableObserver;
import io.reactivex.schedulers.Schedulers;

public class MainActivity extends AppCompatActivity {
    TextView cityField, detailsField, currentTemperatureField, humidity_field, pressure_field, updatedField;
    ArrayList<DisplayClass> weatherList = new ArrayList<>();
    ListAdapter adapter;
    public static final int REQUEST_CODE = 1;
    public static boolean bEdit = false;
    public static boolean bIntentEmpty = true;
    ListView lv;
    private Completable  observable = new Completable() {
        @Override
        protected void subscribeActual(CompletableObserver observer) {

        }
    };
    private WeatherViewModel mWeatherViewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getSupportActionBar().hide();
        setContentView(R.layout.activity_main);

        weatherList.clear();
        mWeatherViewModel = ViewModelProviders.of(this).get(WeatherViewModel.class);

        cityField = (TextView) findViewById(R.id.cityField);
        updatedField = (TextView) findViewById(R.id.updateField);
        detailsField = (TextView) findViewById(R.id.detailField);
        currentTemperatureField = (TextView) findViewById(R.id.tempField);
        humidity_field = (TextView) findViewById(R.id.humidField);
        pressure_field = (TextView) findViewById(R.id.pressureField);
        lv = (ListView) findViewById(R.id.weatherLV);
        Button addButton = (Button) findViewById(R.id.AddBTN);

        addButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                bEdit = false;
                bIntentEmpty = false;
                Intent intent = new Intent(v.getContext(), Main3Activity.class);
                startActivityForResult(intent, REQUEST_CODE);
            }
        });

        adapter = new ListAdapter(this, weatherList);
        lv.setAdapter(adapter);
        mWeatherViewModel.getmAllWeather().observe(this, new Observer<List<DisplayClass>>() {
            @Override
            public void onChanged(@Nullable List<DisplayClass> displayClasses) {
                adapter.setmWeathers(displayClasses);
            }
        });

    }

    @Override
    protected void onResume() {
        super.onResume();
        mWeatherViewModel.getmAllWeather().observe(this, new Observer<List<DisplayClass>>() {
            @Override
            public void onChanged(@Nullable List<DisplayClass> displayClasses) {
                adapter.setmWeathers(displayClasses);
            }
        });
    }

    @Override
    protected void onPause() {
        super.onPause();
        observable.unsubscribeOn(Schedulers.io());
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if(bEdit == true) {
            DisplayClass receiver = data.getParcelableExtra("UpdateData");
            int position = receiver.getUid();
            weatherList.set(position, receiver);
        }
    }

    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                NavUtils.navigateUpFromSameTask(this);
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }
}

File: app/src/main/java/com/example/weatherapp1/WeatherList/MainActivity.kt
package com.example.weatherapp1.WeatherList

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.content.Intent
import android.os.Bundle
import android.support.v4.app.NavUtils
import android.support.v7.app.AppCompatActivity
import android.view.MenuItem
import android.view.View
import android.widget.Button
import android.widget.ListView
import android.widget.TextView

import com.example.weatherapp1.Data.Model.DisplayClass
import com.example.weatherapp1.R
import com.example.weatherapp1.WeatherManipulator.Main3Activity

import java.util.ArrayList

import io.reactivex.Completable
import io.reactivex.CompletableObserver
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.activity_main3.view.*

class MainActivity : AppCompatActivity() {
    private var weatherList = ArrayList<DisplayClass>()
    private lateinit var adapter: ListAdapter
    private val observable = object : Completable() {
        override fun subscribeActual(observer: CompletableObserver) {

        }
    }
    private var mWeatherViewModel: WeatherViewModel? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        supportActionBar!!.hide()
        setContentView(R.layout.activity_main)

        weatherList.clear()
        mWeatherViewModel = ViewModelProviders.of(this).get(WeatherViewModel::class.java!!)

        val cityField = findViewById<View>(R.id.cityField) as? TextView
        val updatedField = findViewById<View>(R.id.updateField) as? TextView
        val detailsField = findViewById<View>(R.id.detailField) as? TextView
        val currentTemperatureField = findViewById<View>(R.id.tempField) as? TextView
        val humidity_field = findViewById<View>(R.id.humidField) as? TextView
        val pressure_field = findViewById<View>(R.id.pressureField) as? TextView
        val lv = findViewById<View>(R.id.weatherLV) as? ListView
        val addButton = findViewById<View>(R.id.AddBTN) as? Button

        if (addButton != null) {
            addButton.setOnClickListener { v ->
                bEdit = false
                bIntentEmpty = false
                val intent = Intent(v.context, Main3Activity::class.java)
                startActivityForResult(intent, REQUEST_CODE)
            }
        }

        adapter = ListAdapter(this, weatherList)
        if (lv != null) {
            lv.adapter = adapter
        }
        mWeatherViewModel!!.getmAllWeather().observe(this, Observer { displayClasses -> displayClasses?.let { adapter.setmWeathers(it) } })

    }

    override fun onResume() {
        super.onResume()
        mWeatherViewModel!!.getmAllWeather().observe(this, Observer { displayClasses -> displayClasses?.let { adapter.setmWeathers(it) } })
    }

    override fun onPause() {
        super.onPause()
        observable.unsubscribeOn(Schedulers.io())
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (bEdit == true) {
            val receiver = data!!.getParcelableExtra<DisplayClass>("UpdateData")
            val position = receiver.uid
            weatherList[position] = receiver
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                NavUtils.navigateUpFromSameTask(this)
                return true
            }
            else -> return super.onOptionsItemSelected(item)
        }
    }

    companion object {
        val REQUEST_CODE = 1
        var bEdit = false
        var bIntentEmpty = true
    }
}
